modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+    }
+
+  log_msg_ack(msg, path_options, AT_PROCESSED);
+  log_msg_unref(msg);
+
+  return TRUE;
+}
+
+static gboolean
+snmpdest_worker_thread_pop_and_insert(SNMPDestDriver *self, GMutex *queue_mutex, LogPathOptions *path_options)
+{
+  LogMessage *msg = log_queue_pop_head_ignore_throttle(self->queue, path_options);
+  g_mutex_unlock(queue_mutex);
+
+  if (!msg)
+    return FALSE;
+
+  if (!snmpdest_worker_insert(self, msg, path_options))
+    {
+      /* error */
+    }
+
+  return TRUE;
+}
+
+static void
+snmpdest_worker_thread(gpointer arg)
+{
+  SNMPDestDriver *self = (SNMPDestDriver *)arg;
+  LogPathOptions path_options = LOG_PATH_OPTIONS_INIT;
+
+  msg_debug ("Worker thread started",
+             evt_tag_str("driver", self->super.super.id),
+             NULL);
+
+  gboolean popped = TRUE; /* the success of the _pop_head() */
+  while (!self->writer_thread_terminate)
+    {
+      g_mutex_lock(self->queue_mutex);
+      int n = log_queue_get_length(self->queue);
+      if (n > 0 || !popped)
+        {
+          /* there are items in the queue or the previous pop was unsuccesful (the msg is not yet present in the queue) */
+          popped = snmpdest_worker_thread_pop_and_insert(self, self->queue_mutex, &path_options);
+          if (!popped)
+            usleep(0);
+
+          continue;
+        }
+
+      /* wait for the signal after the message has been pushed */
+      if (!self->writer_thread_terminate)
+        g_cond_wait(self->writer_thread_wakeup_cond, self->queue_mutex);
+      if (self->writer_thread_terminate)
+        {
+          g_mutex_unlock(self->queue_mutex);
+          break;
+        }
+
+      popped = snmpdest_worker_thread_pop_and_insert(self, self->queue_mutex, &path_options);
+    }
+
+  msg_debug ("Worker thread finished",
+             evt_tag_str("driver", self->super.super.id),
+             NULL);
+
+}
+
+static void
+snmpdest_dd_stop_thread(gpointer arg)
+{
+  SNMPDestDriver *self = (SNMPDestDriver *)arg;
+  self->writer_thread_terminate = TRUE;
+  g_mutex_lock(self->queue_mutex);
+  g_cond_signal(self->writer_thread_wakeup_cond);
+  g_mutex_unlock(self->queue_mutex);
+}
+
+static void
+snmpdest_dd_start_thread(SNMPDestDriver *self)
+{
+  main_loop_create_worker_thread(snmpdest_worker_thread, snmpdest_dd_stop_thread, self, &self->worker_options);
+}
+
+static gchar *
+snmpdest_dd_format_persist_name(SNMPDestDriver *self)
+{
+  static gchar persist_name[1024];
+
+  g_snprintf(persist_name, sizeof(persist_name), "snmpdest(%s,%u)", self->host, self->port);
+  return persist_name;
+}
+
+static gchar *
+snmpdest_dd_format_stats_instance(SNMPDestDriver *self)
+{
+  static gchar persist_name[1024];
+
+  g_snprintf(persist_name, sizeof(persist_name), "snmpdest,%s,%u", self->host, self->port);
+  return persist_name;
+}
+
+static gboolean
+snmpdest_dd_session_init(SNMPDestDriver *self)
+{
+  /* SNMP session setup */
+  memset(&self->session, 0, sizeof(self->session));
+
+  putenv(strdup("POSIXLY_CORRECT=1"));
```

Comment:
Do we need the strdup here? This causes memleak.

Reply:
I guess we need strdup here. This environmental variable can be used later in the code. I just add a pointer to a stack-allocated string it will cause problems outside the scope of this init function.

Reply:
based on our discussion: the string literal is allocated in the program segment so don't need to use strdup here. removed.